### DISCUSSION TITLE
What about HTML Streaming?

### DISCUSSION THREAD

**Discussion by ghost** [nanojsx/nano#97] [General]
Nowadays SSR is being made with async rendering which is a way rendering a page on the server waiting for data to come or finish the page logic & generate HTML, but this can inefficient & blocking because if a data latency is high or if a data to render is huge it's not good to wait to send static files & render static parts.

So There is a mechanism called HTML streaming which allows to render as we fetch a data means we don't need to block rendering waiting for a response from an API, we can start rendering the static parts & show loading spinner for dynamic data.

To see it in action test these web site that is built with solid.js, just open it, go to dev tools & set network throttlling. 

https://hackernews.ryansolid.workers.dev/

https://github.com/solidjs/solid-start/tree/main/examples/hackernews-ts

I like to say that you are nanojsx is really awesome. See how fast the Deno docs site is which is built with Nanossr.
 https://pagespeed.web.dev/report?url=https%3A%2F%2Fdoc.deno.land%2F

**Comment by yandeu** [comment]
Wow. I did not know that doc.deno.land was build using Nano JSX.

What you did with solidjs is also totally possible with Nano JSX.  
Nano JSX is 100% isomorphic, means it work on the server and in the browser.

**Reply by yandeu** [reply]
The term "HTML Streaming" doesn't really exist. You probably mean progressive rendering or partial rendering. It is up to your server to send the html response, not NanoJSX. You can configure your setup however you want to send the initial response to the client, and fetch more things later.

**Reply by ghost** [reply]
Just see this article from cloudflare that explains about TransformStream.

https://developers.cloudflare.com/workers/learning/using-streams

https://m.youtube.com/watch?v=8hlo9uPrpr4

https://m.youtube.com/watch?v=G9PpImUEeUA

**Reply by jrson83** [reply]
Nano JSX uses `Nano.renderSSR()` function which returns an HTML string, like as an example React does with [renderToString](https://reactjs.org/docs/react-dom-server.html#rendertostring).
It does not have a [React like function](https://reactjs.org/docs/react-dom-server.html#rendertonodestream) `renderToNodeStream()`. Here is a guide on [Streaming Server-Side Rendering](https://mxstbr.com/thoughts/streaming-ssr/) with React, on which you see that the `renderToNodeStream` function is different then `renderToString`.

If you check the [solid/js docs](https://www.solidjs.com/guides/server) and also the example [solid/js app you posted](https://github.com/solidjs/solid-start/blob/main/examples/hackernews-ts/src/entry-server.tsx), you can see it also supports `renderToString` and additionally `renderToNodeStream`.

So I don't think that Nano supports streams at the moment. However you could use `fetch()` with `Suspense()`, if you want to display a loading spinner ([example](https://nanojsx.io/examples.html#fetchSuspenseIsomorphic)).

For your hackernews example, even if I turn on network throttlling, I don't see any loading/spinner since the data is loading directly.

**Reply by yandeu** [reply]
Here is a stream example using nanojsx:
https://github.com/nanojsx/progressive-rendering-example

**Reply by ghost** [reply]
Thanks. I am saying this because renderToString is something problematic, it can block rendering & take some seconds to get to the user.

I am just a fan of Deno & Deno deploy, Although it's in beta 3. Today I am able to Deploy my simple site using Deno, Oak, Nano_Jsx, Tailwind css. This App makes some API calls to get a data, but it's fast enough to me & I am happy.

Just See The Site's Page Speed scores.

![Screenshot_20220219-160753_Chrome.jpg](https://user-images.githubusercontent.com/96905008/154815419-9b5c7a2b-b6c9-4482-8955-fca374c23b84.jpg)

![Screenshot_20220219-160723_Chrome.jpg](https://user-images.githubusercontent.com/96905008/154815424-e75ce49d-788d-4a2f-b8b8-dfe95154f93c.jpg)

**Reply by yandeu** [reply]
Looks great!

renderToString (renderSSR in Nano JSX) should be very very fast.

You could also put a CDN in front of your app.

You could also try to inline your css in the html document to increase FirstContentfulPaint.

**Reply by yandeu** [reply]
I get exactly the same results (0.8s) with Nano JSX and CloudFront CDN.
